### PR TITLE
fix: Add support for the new snapshot API

### LINKED ISSUE #12960: Legacy accessibility snapshots are no longer supported on  Xcode 11 beat 4 /iOS 13
latest   appium/WebDriverAgent

launch WebDriverAgentRunner 

http://localhost:8100/source

WebDriverAgentRunner-Runner[23050:265535] Internal error: Error Domain=XCTDaemonErrorDomain Code=39 "Legacy accessibility snapshots are no longer supported" UserInfo={NSLocalizedDescription=Legacy accessibility snapshots are no longer supported}

### PR DESCRIPTION
**Pull request by umutuzgur**
Fixes https://github.com/appium/appium/issues/12960

### CODE REVIEW

File: WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m [hunk 1]
```diff
@@ -147,6 +110,76 @@ - (nullable XCElementSnapshot *)fb_snapshotWithAttributes {
   return snapshotWithAttributes;
 }
 
++ (BOOL) fb_newSnapshotAPIIsSupported
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m]: it is a good practice to put is as name prefix: fb_is...

File: WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m [hunk 2]
```diff
@@ -147,6 +110,76 @@ - (nullable XCElementSnapshot *)fb_snapshotWithAttributes {
...
+{
+  static dispatch_once_t newSnapshotIsSupported;
+  static BOOL result;
+  dispatch_once(&newSnapshotIsSupported, ^{
+    result = [(NSObject *)[FBXCTestDaemonsProxy testRunnerProxy] respondsToSelector:@selector(_XCT_requestSnapshotForElement:attributes:parameters:reply:)];
+  });
+  return result;
+}
+
+- (void) fb_requestSnapshot:(XCAccessibilityElement *)accessibilityElement reply:(void (^)(XCElementSnapshot *, NSError *))block
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m]: the common style does not have space after return type declaration

File: WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m [hunk 3]
```diff
@@ -147,6 +110,76 @@ - (nullable XCElementSnapshot *)fb_snapshotWithAttributes {
...
+                                    reply:block];
+  } else {
+    [proxy _XCT_snapshotForElement:self.lastSnapshot.accessibilityElement
+                        attributes:axAttributes
+                        parameters:defaultParameters
+                             reply:block];
+  }
+}
+
+- (NSArray *) fb_createAXAttributes: (BOOL)asNumber
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m]: no need for space after the colon

File: WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m [hunk 4]
```diff
@@ -147,6 +110,76 @@ - (nullable XCElementSnapshot *)fb_snapshotWithAttributes {
...
+  ];
+  
+  SEL attributesForElementSnapshotKeyPathsSelector = [XCElementSnapshot fb_attributesForElementSnapshotKeyPathsSelector];
+  NSSet *attributes = (nil == attributesForElementSnapshotKeyPathsSelector) ? nil
+  : [XCElementSnapshot performSelector:attributesForElementSnapshotKeyPathsSelector withObject:propertyNames];
+  if (attributes == nil) {
+    @throw [NSException exceptionWithName:@"AttributesEmpty" reason:@"Couldn't build the attributes " userInfo:nil];
+  }
+  if (asNumber) {
+    NSArray *axAttributes = XCAXAccessibilityAttributesForStringAttributes(attributes);
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m]: I'd rather get mutable array here and then just return a copy of it at the end

File: WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m [hunk 5]
```diff
@@ -147,6 +110,76 @@ - (nullable XCElementSnapshot *)fb_snapshotWithAttributes {
...
+  });
+  return result;
+}
+
+- (void) fb_requestSnapshot:(XCAccessibilityElement *)accessibilityElement reply:(void (^)(XCElementSnapshot *, NSError *))block
+{
+  static NSDictionary *defaultParameters;
+  static dispatch_once_t initializeAttributesAndParametersToken;
+  static NSArray *axAttributes;
+  BOOL useNewSnapshotAPI = [XCUIElement fb_newSnapshotAPIIsSupported];
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m]: some explanation comment here might be useful

File: PrivateHeaders/XCTest/XCTestManager_ManagerInterface-Protocol.h [hunk 6]
```diff
@@ -27,6 +27,7 @@
 - (void)_XCT_setAttribute:(NSNumber *)arg1 value:(id)arg2 element:(XCAccessibilityElement *)arg3 reply:(void (^)(BOOL, NSError *))arg4;
 - (void)_XCT_fetchAttributesForElement:(XCAccessibilityElement *)arg1 attributes:(NSArray *)arg2 reply:(void (^)(NSDictionary *, NSError *))arg3;
 - (void)_XCT_snapshotForElement:(XCAccessibilityElement *)arg1 attributes:(NSArray *)arg2 parameters:(NSDictionary *)arg3 reply:(void (^)(XCElementSnapshot *, NSError *))arg4;
```
Comment (mykola-mokhnach) [inline PrivateHeaders/XCTest/XCTestManager_ManagerInterface-Protocol.h:30]: Can we add a comment here in which iOS version the method has been deprecated?

File: WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m [hunk 7]
```diff
@@ -147,6 +110,76 @@ - (nullable XCElementSnapshot *)fb_snapshotWithAttributes {
...
+                        parameters:defaultParameters
+                             reply:block];
+  }
+}
+
+- (NSArray *) fb_createAXAttributes: (BOOL)asNumber
+{
+  // Names of the properties to load. There won't be lazy loading for missing properties,
+  // thus missing properties will lead to wrong results
+  NSArray<NSString *> *propertyNames = @[
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m:152]: is probably makes sense to make this array static and fill it with values inside dispatch_once block

File: WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m [hunk 8]
```diff
@@ -255,8 +288,8 @@ - (NSData *)fb_screenshotWithError:(NSError **)error
           || (appFrame.size.width > appFrame.size.height && parentWindowFrame.size.width > parentWindowFrame.size.height)
           || (appFrame.size.width < appFrame.size.height && parentWindowFrame.size.width < parentWindowFrame.size.height)) {
         CGPoint fixedOrigin = orientation == UIInterfaceOrientationLandscapeLeft ?
-          CGPointMake(appFrame.size.height - elementRect.origin.y - elementRect.size.height, elementRect.origin.x) :
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m]: I think it makes sense to keep the previous indentation

File: WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m [hunk 9]
```diff
@@ -235,7 +268,7 @@ - (NSData *)fb_screenshotWithError:(NSError **)error
     }
     return nil;
   }
-
+
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m]: not needed %)

File: WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m (hunk 7 above)
Comment (umutuzgur) [reply WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m:152]: No need to do extra optimisation here.  `fb_createAXAttributes` will be called once anyways by fb_requestSnapshot. I don't like using dispatch_once a lot. It is a bad practice

File: WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m (hunk 8 above)
Comment (umutuzgur) [reply WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m]: I can revert this part but this was auto indentation from XCode. Just FYI

File: WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m (hunk 7 above)
Comment (mykola-mokhnach) [reply WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m:152]: with the current approach we create a new array object every time we enter this method (e.g take a snapshot). Putting it into static will save some milliseconds on each call

Comment (umutuzgur) [reply WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m:152]: I can optimise that later on but this is seriously not a big issue

**Review by mykola-mokhnach — APPROVED** [review]

**Comment by mykola-mokhnach** [comment]
Please merge after functional test pass

File: PrivateHeaders/XCTest/XCTestManager_ManagerInterface-Protocol.h [hunk 10]
```diff
@@ -43,4 +44,7 @@
 // Available since Xcode9
 - (void)_XCT_requestScreenshotOfScreenWithID:(unsigned int)arg1 withRect:(struct CGRect)arg2 uti:(NSString *)arg3 compressionQuality:(double)arg4 withReply:(void (^)(NSData *, NSError *))arg5;
 - (void)_XCT_requestScreenshotOfScreenWithID:(unsigned int)arg1 withRect:(struct CGRect)arg2 withReply:(void (^)(NSData *, NSError *))arg3;
+
+// Available since Xcode11
```
Comment (KazuCocoa) [inline PrivateHeaders/XCTest/XCTestManager_ManagerInterface-Protocol.h:48]: [nits]
This was available since Xcode 9.0, according to header dump

**Review by KazuCocoa — APPROVED** [review]
:tada:
I ran some commands with this change such as find element/s, source and send keys

**Comment by KazuCocoa** [comment]
Out of this PR's issue, but let me leave a note.
iOS 13 returns wrong coordinates for elements. `element.click` clicks the wrong place. (It taps a bit over place for the element)
I wondered status-bar or something affected it... (I haven't checked iOS 12 x Xcode 11 to ensure it happens only on iOS 13)

**Comment by umutuzgur** [comment]
@KazuCocoa You mean the page source is wrong? We did a page source diff and it was the same as the old method

**Comment by KazuCocoa** [comment]
Yes. That was iOS 13 issue, bot Xcode 11 since Xcode 11 x iOs 12.1 was correct.

https://gist.github.com/KazuCocoa/e5ed3b266da87c68882379ba2c02c2ef

~~Will check with Xcode 11 beta 6~~

**Comment by KazuCocoa** [comment]
settings app could return correct value.
Maybe we should update our test app

**Comment by KazuCocoa** [comment]
Once we apply the lates ui-catalog by Apple, ^ did not happen.
https://github.com/appium/ios-uicatalog/pull/15

If someone reported this kind of issue, we should guide them to update their production app for iOS 13.

**Pull request merged**